src/OpenTracing/Mock/MockSpan.cs

```diff
@@ -0,0 +1,306 @@
+                CheckForFinished("Adding baggage [{0}:{1}] to already finished span.", key, value);
+                _context = _context.WithBaggageItem(key, value);
+                return this;
+            }
+        }
+
+        public string GetBaggageItem(string key)
+        {
+            lock (_lock)
+            {
+                return _context.GetBaggageItem(key);
+            }
+        }
+
+        public void Finish()
+        {
+            Finish(DateTimeOffset.UtcNow);
+        }
+
+        public void Finish(DateTimeOffset finishTimestamp)
+        {
+            lock (_lock)
+            {
+                CheckForFinished("Tried to finish already finished span");
+                _finishTimestamp = finishTimestamp;
+                _mockTracer.AppendFinishedSpan(this);
+                _finished = true;
+            }
+        }
+
+        private static MockSpanContext FindPreferredParentRef(IList<Reference> references)
+        {
+            if (!references.Any())
+                return null;
+
+            // return the context of the parent, if applicable
+            foreach (var reference in references)
+            {
+                if (OpenTracing.References.ChildOf.Equals(reference.ReferenceType))
+                    return reference.Context;
+            }
+
+            // otherwise, return the context of the first reference
+            return references.First().Context;
+        }
+
+        private static Dictionary<string, string> MergeBaggages(IList<Reference> references)
+        {
+            var baggage = new Dictionary<string, string>();
+            foreach (var reference in references)
+            {
+                if (reference.Context.GetBaggageItems() != null)
+                {
+                    foreach (var bagItem in reference.Context.GetBaggageItems())
+                    {
+                        baggage[bagItem.Key] = bagItem.Value;
+                    }
+                }
+            }
+
+            return baggage;
+        }
+
+        private void CheckForFinished(string format, params object[] args)
+        {
+            if (_finished)
+            {
+                var ex = new InvalidOperationException(string.Format(format, args));
+                _errors.Add(ex);
+                throw ex;
+            }
+        }
+
+        public override string ToString()
+        {
+            return $"TraceId: {_context.TraceId}, SpanId: {_context.SpanId}, ParentId: {ParentId}, OperationName: {OperationName}";
+        }
+
+        public sealed class LogEntry
+        {
+            public DateTimeOffset Timestamp { get; }
+
+            public IReadOnlyDictionary<string, object> Fields { get; }
+
+            public LogEntry(DateTimeOffset timestamp, IDictionary<string, object> fields)
+            {
+                Timestamp = timestamp;
+                Fields = new ReadOnlyDictionary<string, object>(fields);
+            }
+        }
+
+        public struct Reference
```

Comment:
[nit] you should generally prefer to implement IEquatable anyway, especially if used in a distinct collection. It'll help you avoid subtle bugs and other-dev-proof your code (e.g. right now it's not very clear that we have constraints on this type properly implementing equality. I'm also inclined to believe (e.g. I wrote it then second guessed myself but don't want to measure right now) that the default equality comparer will result in boxing for a struct, unless you've implemented IEquatable.

Reply:
I checked that ( e.g. https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/how-to-define-value-equality-for-a-type ) and it's true that the default equality comparison for structs is slow (as it uses reflection). I'll change it.